### PR TITLE
Rustup to "rustc 1.59.0-nightly (78fd0f633 2021-12-29)"

### PR DESCRIPTION
**Pull request by bjorn3**
There were a couple of conflicts I had to resolve. If you want I can do a PR to sync it the other way around too.

### CODE REVIEW

File: src/base.rs [hunk 1]
```diff
@@ -135,40 +133,3 @@ pub fn compile_codegen_unit<'tcx>(tcx: TyCtxt<'tcx>, cgu_name: Symbol) -> (Modul
...
-    let global = context.new_global(None, GlobalKind::Exported, typ, name);
-    global.global_set_initializer(&compressed);
-    global.set_link_section(section_name);
-
-    // Also generate a .section directive to force no
-    // flags, at least for ELF outputs, so that the
-    // metadata doesn't get loaded into memory.
-    let directive = format!(".section {}", section_name);
-    context.add_top_level_asm(None, &directive);
-}
```
Comment (bjorn3) [inline src/base.rs:174]: cg_ssa now has builtin support for writing the dylib metadata object.

**Comment by bjorn3** [comment]
Is this pre-existing?

```
failures:

---- slice::test_is_sorted stdout ----
---- slice::test_is_sorted stderr ----
thread 'main' panicked at 'attempt to create unaligned or null slice', /home/runner/work/rustc_codegen_gcc/rustc_codegen_gcc/build_sysroot/sysroot_src/library/core/src/slice/raw.rs:141:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    slice::test_is_sorted

test result: FAILED. 1381 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 16.79s

error: test failed, to rerun pass '--lib'
```

**Comment by antoyo** [comment]
libcore's test used to all pass. Maybe that's a new test that found a new bug as [it previously only ran 1337 tests](https://github.com/rust-lang/rustc_codegen_gcc/runs/4543304957?check_suite_focus=true#step:15:2625) and this PR runs 1382 tests.

**Comment by bjorn3** [comment]
Shall I disable it for now?

**Comment by antoyo** [comment]
Yes!

**Comment by antoyo** [comment]
Thanks! Is this ready to be reviewed and merged?

**Comment by bjorn3** [comment]
Yes it is.

File: patches/0028-core-Disable-long-running-tests.patch [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+diff --git a/library/core/tests/slice.rs b/library/core/tests/slice.rs
+index 2c8f00a..44847ee 100644
+--- a/library/core/tests/slice.rs
++++ b/library/core/tests/slice.rs
+@@ -2332,7 +2332,8 @@ macro_rules! empty_max_mut {
+     };
+ }
+ 
++/*
```
Comment (antoyo) [inline patches/0028-core-Disable-long-running-tests.patch:18]: Should this be enabled back?
Not sure why this was disabled.

**Review by antoyo — APPROVED** [review]

File: patches/0028-core-Disable-long-running-tests.patch (hunk 2 above)
Comment (bjorn3) [reply patches/0028-core-Disable-long-running-tests.patch:18]: These tests only finish within reasonable time when optimizations are enabled. The libcore tests are currently tested with optimizations disabled.

Comment (antoyo) [reply patches/0028-core-Disable-long-running-tests.patch:18]: Oh wow, I did not notice that. I guess those are new tests. Let's keep them disabled, then.

**Pull request merged**